pkg/rhcos/openstack.go

```diff
@@ -49,10 +49,16 @@ func GenerateOpenStackImageName(rhcosImage, infraID string) (imageName string, i
 	// should be created by the installer with the universal name "<infraID>-rhcos". Otherwise, it means
 	// that we are given the name of the pre-created Glance image, which the installer should use for node
 	// provisioning.
-	_, err := url.ParseRequestURI(rhcosImage)
+	// We support only 'http(s)' and 'file' schemes, URLs with other schemes are considered as Glance image
+	// names.
+	url, err := url.Parse(rhcosImage)
 	if err != nil {
 		return rhcosImage, false
 	}
 
+	if url.Scheme != "http" && url.Scheme != "https" && url.Scheme != "file" {
+		return rhcosImage, false
+	}
+
```

Comment:
technically one could provide a ftp URI and that would still work, but now providing something like that means it's going to try to assume that as glance image id.

I think there needs to be better validation here in terms of whats allowed and what's not.

Reply:
Now  the installation fails if the user provides an ftp or any other non-http(s) url, because we will try to download the file using http protocol. This is wrong behavior.
With this patch I limit the supported schemes to http(s) and file, as it is said in the documentation. URLs with other schemes will be considered as Glance image names.

Reply:
but this function will return result for ftp as if that's an glance image though still.. will is incorrect.

we should proabably have validation to restrict the scheme to these values or empty for glance image.

Reply:
In accordance with the documentation and as originally planned:
1. A valid `http(s)` URL translates to the address from where we're going to download the file.
2. A valid `file` URL translates to the image file location on the local file system.
3. All other strings are translated to the Glance image name, even if they are `ftp` or `s3` URLs.

As an alternative solution, I can offer to limit the schemes and return an error if the scheme is not http or file. That is:
1. A valid `http(s)` URL translates to the address from where we're going to download the file.
2. A valid `file` URL translates to the image file location on the local file system.
3. Any other URL-like string leads to the error: "Unsupported URL scheme".
3. All other strings are translated to the Glance image name.